// material-ui/Hidden/types.js
'use strict';

var _react = require('react');

var babelPluginFlowReactPropTypes_proptype_Breakpoint = require('../styles/breakpoints').babelPluginFlowReactPropTypes_proptype_Breakpoint || require('prop-types').any;

if (typeof exports !== 'undefined') Object.defineProperty(exports, 'babelPluginFlowReactPropTypes_proptype_HiddenProps', {
  value: require('prop-types').shape({
    children: require('prop-types').any,
    className: require('prop-types').string,
    only: require('prop-types').oneOfType([babelPluginFlowReactPropTypes_proptype_Breakpoint, require('prop-types').arrayOf(babelPluginFlowReactPropTypes_proptype_Breakpoint)]),
    xsUp: require('prop-types').bool,
    smUp: require('prop-types').bool,
    mdUp: require('prop-types').bool,
    lgUp: require('prop-types').bool,
    xlUp: require('prop-types').bool,
    xsDown: require('prop-types').bool,
    smDown: require('prop-types').bool,
    mdDown: require('prop-types').bool,
    lgDown: require('prop-types').bool,
    xlDown: require('prop-types').bool
  })
});
